### DISCUSSION TITLE
Automatic change of base branch dismisses PR reviews

### DISCUSSION THREAD

**Discussion by ffissore** [community/community#57513] [Pull Requests]
### Select Topic Area

Bug

### Body

In order to create PRs focused on one particular kind of change, and thus easier to review, we are used to create what we call "PRs trains":
* PR from branch `A` wants to merge into `main`, PR from branch `B` wants to merge into `A`, and so on
* reviewers review each PR
* PR from branch `A` is merged, we wait until deployment goes well and all is green
* github automatically changes the base branch of PR from branch `B`
* PR from branch `B` is merged

Since this morning this flawless flow is no longer possible: my colleagues reviewed my PRs but when the first one was merged I got 
![image](https://github.com/community/community/assets/510465/00519acb-b72e-4866-97fe-cae31bc72170)
and I had to ask my colleagues to reapprove.

Can you revert this change?

**Comment by stanfeldman** [comment]
This very annoying change breaks a typical workflow when you create incremental staggered PRs for a feature. We've got extra protection for a very unlikely scenario that inhibits some legit flows that are more common.
[There are other people raising similar concerns](https://github.com/orgs/community/discussions/57944).

**Comment by ffissore** [comment]
Hello github?

**Comment by ffissore** [comment]
I opened a support issue with github. The workaround is to disable/enable a couple of options in the branch protection rule

> As a workaround, some customers have disabled the rule Dismiss stale pull request approvals when new commits are pushed and instead have enabled Require approval of the most recent reviewable push. These rules are slightly different - I'll describe them here:
>
> Dismiss stale pull request approvals when new commits are pushed
>
> Optionally, you can choose to dismiss stale pull request approvals when commits are pushed that affect the diff in the pull request. GitHub records the state of the diff at the point when a pull request is approved. This state represents the set of changes that the reviewer approved. If the diff changes from this state (for example, because a contributor pushes new changes to the pull request branch or clicks Update branch, or because a related pull request is merged into the target branch), the approving review is dismissed as stale, and the pull request cannot be merged until someone approves the work again
>
> Require approval of the most recent reviewable push
>
> Optionally, you can require that the most recent reviewable push must be approved by someone other than the person who pushed it. This means at least one other authorized reviewer has approved any changes. For example, the "last reviewer" can check that the latest set of changes incorporates feedback from other reviews, and does not add new, unreviewed content.

**Comment by mariux** [comment]
We experience the same issue but we can not use the provided workaround that @ffissore shared. 

What is the expected timeline for a fix here as this blocks our workflows and frustrates our team ;) 

Thanks for GitHub and the cewl features please make it work as awesome as before ;)

**Comment by alec-oms** [comment]
Getting the same problem here and is reducing team velocity and productivity.

Tried the work around and didn't work either. PR's ended up in a limbo state where review checks were complete, but still could not be merged without reverting the workaround, then closing and re-opening the PR.

Keen to find out when this will be fixed, or if there is another process/work around to achieve the same PR chaining approach.

**Comment by ffissore** [comment]
GH confirmed the suggestion above didn't aim at restoring the previous behaviour. They also said that at the moment the only way to have something similar is to disable `Dismiss stale pull request approvals when new commits are pushed`, which is a security concern and thus not applicable.
I'm afraid it's just broken.
Not sure who they talked to before deciding such a change, but I'd love to know what PR flows these github customers implement, because ours is broken and we have to harass each other more often asking for re-approvals

**Comment by joshuakeel** [comment]
Totally agree, this is affecting our team as well, and is really making our workflows more clunky.

**Comment by ZeeshanMukhtar1** [comment]


**Reply by joshuakeel** [reply]
This setting is under the "General" section, and not "Branches", at least for me.

In addition, this setting is not turned on, and we are still having problems.

**Reply by pbs-jmo** [reply]
This isn't actually a solution. We should still be able to auto-prune branches and also keep the old behavior which was to not dismiss reviews as "stale" due to that happening.

**Comment by aaditmshah** [comment]
This is a very unfortunate regression. In my team, we introduced a new workflow. The person who reviews PRs is also responsible for merging them. They should merge the PRs as and when they are done reviewing them. That way, they don't need to re-approve PRs. Also, when PRs are marked as ready for review they are also ready to be merged. Thus, the reviewer can do their job asynchronously from other developers.

**Reply by aaditmshah** [reply]
![State Chart of Pull Requests](https://github.com/community/community/assets/813511/64498c9c-40dd-47d6-8c9c-2d139afbd47a)

**Reply by ffissore** [reply]
This is an interesting approach, thx @aaditmshah I'll give it a try

**Reply by davidkocnar** [reply]
What about chained PRs? You cannot review them as soon as their base is not `develop` and the base branch cannot be approved/merged.

**Reply by aaditmshah** [reply]
@davidkocnar For chained PRs you can use fast-forward merges. For example, suppose you have the following commit history.

```
A --- B --- C --- D --- E --- F --- G --- H --- I --- J --- K --- L --- M --- N --- O
            ^                 ^                 ^                 ^                 ^
            |                 |                 |                 |                 |
          main             develop           fix/foo         feature/bar       feature/baz
```

We have the following four PRs.

1. `develop` to `main`
2. `fix/foo` to `develop`
3. `feature/bar` to `fix/foo`
4. `feature/baz` to `feature/bar`

The reviewer can review and approve all the PRs using the GitHub UI. But when merging, the reviewer can merge the branches locally using fast-forward merges, and push the changes to GitHub.

```shell
git fetch # ensure that you have the latest changes locally
git checkout develop # you can run `git merge --ff-only origin/develop` to get the latest changes if required
git merge --ff-only fix/foo # merge the changes from `fix/foo` into `develop`
git push # you can even push to protected branches using fast-forward merges
git merge --ff-only feature/bar # now you can merge the changes from `feature/bar` into `develop`
git push # when using fast-forward merges you don't need to review the PRs again
git merge --ff-only feature/baz # finally you can merge the changes from `feature/baz` into `develop`
git push # github automatically marks the corresponding PRs as merged when using fast-forward merges
git checkout main # now we want to merge `develop` into `main`
git merge --ff-only develop # fast-forward merges can be done in any order
git push # everything can be done locally, without interacting with the github ui
git remote prune origin # after you are done remember to clean up the branches in your remote
```

If you use fast-forward merges frequently, like I do, then you might want to use the following git settings.

```shell
git config --global pull.rebase false
git config --global pull.ff only
git config --global merge.ff only
```

If you enable these settings then you don't need to pass the `--ff-only` flag manually.

**Comment by pbs-jmo** [comment]
This also breaks a common workflow for us at my company. This is pretty inconvenient.

**Comment by ChristopherGerdes** [comment]
This is also breaking our teams workflow.  are there any plans to fix this?

I tried the suggestion of disabling "dismiss stale pull request approvals when new commits are pushed" and enable "require approval of most recent reviewable push" and it did not work.

**Comment by Geri-Borbas-Shape** [comment]
Hey Github, as a workaround: can't be re-approval requests **automatically approved** if there is no code change compared to the previous approval?

**Comment by ChristopherGerdes** [comment]
as an update to this thread - I filed an issue with GitHub and they told me it was by design.  They said they would forward the request to their product team but I may not hear back.

sounds like this is unlikely to get fixed unfortunately.  Too bad since this breaks workflows for a lot of people.

**Comment by tfrokt** [comment]
same here, really annoying "feature".

**Comment by max0700** [comment]
We experience the same issue but we can not use the provided workaround that @ffissore shared.

What is the expected timeline for a fix here as this blocks our workflows and frustrates our team ;)

Thanks for GitHub and the cewl features please make it work as awesome as before ;)

**Comment by fcano-ut** [comment]
The GitHub interface is now inconsistent with itself. I'm getting PRs dismissed, ask reviewers to review again, and when they do they see in the interface that there are no changes to review. So I'm just asking them to approve an empty set of commits. I don't think it makes a lot of sense...

**Comment by Geri-Borbas-Shape** [comment]
> So I'm just asking them to approve an empty set of commits.

Exactly ☝️, even only changing that could be a huge improvement (see https://github.com/orgs/community/discussions/57513#discussioncomment-7680274).

**Comment by markalanevans** [comment]
Same Issue.  How do you stack PR's so that they can easily be merged and reviewed.

**Comment by ben-cronometer** [comment]
Same here.. Please, GitHub! Fix this! Why make such pain???

**Comment by joehorsnell** [comment]
Still an issue two and a half years later GitHub. This completely breaks stacked PR workflows - please can someone provide an update on when this will be fixed.

**Comment by Fyko** [comment]
Here's to hoping this is resolved by https://github.github.com/gh-stack/

**Reply by alexkarcher** [reply]
This is in fact made much worse by stacked PRs. Same behavior, but now it's a formal GH feature that has a lot of us trying the same workflow. When I approve each PR in a stack it dismisses the approval from all the above PRs.

**Reply by Jossec101** [reply]
Gh stack has solved this issue for me

**Comment by ffissore** [comment]
PR stacks are far from what we could do 3 years ago:
* reviews are dismissed
* when stacked PRs get rebased
  * they get rebased and there's no way to tell github to merge instead (this slows down reviews)
  * verified commits become unverified: this is a blocker for us as verified commits are part of our branch protection rules

Still far

**Reply by gnikolaropoulos** [reply]
I have never yet managed to merge a stack using the new UI. I will try with the cli but the UI is really unusable. Merges keep spinning and they never finish with no error message or any indication or feedback of what is going wrong. It feels really broken even for a preview feature.

**Reply by leettaylor** [reply]
@gnikolaropoulos You're not the only one with that issue (https://github.com/github/gh-stack/discussions/256)
